cmd/disk_wipe.go

```diff
@@ -74,14 +76,27 @@ func init() {
 
 			// Pick the most appropriate wipe based on the disk type and/or features supported
 			var wiper actions.DriveWiper
-			// nolint:gocritic // will have more cases soon, remove nolint then
 			switch drive.Protocol {
 			case "nvme":
 				wiper = utils.NewNvmeCmd(verbose)
+			case "sata":
+				// Lets figure out if the drive supports TRIM
+				var trim bool
+				for _, cap := range drive.Capabilities {
+					if strings.HasPrefix(cap.Description, "Data Set Management TRIM supported") {
```

Comment:
does this work across different drives vendors/models?

Reply:
Hmm, I'm seeing some apparently-conflicting info on this -- on a Micron SSD in an n2.xlarge:

```
$ hdparm -I /dev/sda | grep -i trim
	   *	Data Set Management TRIM supported (limit 8 blocks)
	   *	Deterministic read ZEROs after TRIM
$ cat /sys/block/sda/queue/discard_zeroes_data 
0
```

I'm not sure offhand what exactly the difference is between what the kernel reports in `discard_zeroes_data` and what hdparm reports as "Deterministic read ZEROs after TRIM" -- anyone happen to know?

Reply:
Hah: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=48920ff2a5a940cd07d12cc79e4a2c75f1185aee

> Will always return 0.  Don't rely on any specific behavior for discards, and don't read this file.

Tracing through some kernel codepaths, it looks like maybe we should be trying to use the `BLKZEROOUT` ioctl instead...which is available via `blkdiscard --zeroout`, but it doesn't look like ironlib currently passes that flag?

Reply:
> Hmm, I'm seeing some apparently-conflicting info on this -- on a Micron SSD in an n2.xlarge:
> 
> ```
> $ hdparm -I /dev/sda | grep -i trim
> 	   *	Data Set Management TRIM supported (limit 8 blocks)
> 	   *	Deterministic read ZEROs after TRIM
> $ cat /sys/block/sda/queue/discard_zeroes_data 
> 0
> ```
> 
> I'm not sure offhand what exactly the difference is between what the kernel reports in `discard_zeroes_data` and what hdparm reports as "Deterministic read ZEROs after TRIM" -- anyone happen to know?


> Hah: [git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit?id=48920ff2a5a940cd07d12cc79e4a2c75f1185aee](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=48920ff2a5a940cd07d12cc79e4a2c75f1185aee)
> 
> > Will always return 0.  Don't rely on any specific behavior for discards, and don't read this file.
> 
> Tracing through some kernel codepaths, it looks like maybe we should be trying to use the `BLKZEROOUT` ioctl instead...which is available via `blkdiscard --zeroout`, but it doesn't look like ironlib currently passes that flag?

Looks like discard_zeros_data was just internal use to me. That patch series also adds a bunch of code handling `REQ_OP_WRITE_ZEROES` and keeps `REQ_OP_DISCARD` around:

```
 	if (op == REQ_OP_DISCARD)
 		special_cmd_max_sectors = q->limits.max_discard_sectors;
+	else if (op == REQ_OP_WRITE_ZEROES)
```

From the blkdiscard manpage:
```
       -z, --zeroout
           Zero-fill rather than discard.
```
which is not what we want :smile: (tried running this and its definitely taking a long time to come finish and is in `D+` state. fwiw hdparm gets the data from `SG_IO` ioctl:
```
ioctl(3, SG_IO, {interface_id='S', dxfer_direction=SG_DXFER_FROM_DEV, cmd_len=16, cmdp="\x85\x09\x0e\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\xe0\x2f\x00", mx_sb_len=32, iovec_count=0, dxfer_len=512, timeout=15000, flags=0, dxferp="\x01\x00\x00\x00\x00\x00\x14\x00\x08\x00\x00\x00\x00\x00\x02\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"..., status=0, masked_status=0, msg_status=0, sb_len_wr=0, sbp="", host_status=0, driver_status=0, resid=0, duration=0, info=0}) = 0                                                   
```